File: src/main/java/com/LeaseManager/Entity/Contract.java
package com.LeaseManager.Entity;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

/**
 * Договор лизинга
 */
@Entity
@Table(name = "contracts")
@Builder
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Contract {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "contract_number", nullable = false, unique = true, length = 50)
    private String contractNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "client_id", nullable = false, foreignKey = @ForeignKey(name = "fk_contract_client"))
    private Client client;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "equipment_id", nullable = false, foreignKey = @ForeignKey(name = "fk_contract_equipment"))
    private Equipment equipment;

    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;

    @Column(name = "end_date", nullable = false)
    private LocalDate endDate;

    @Column(name = "total_amount", nullable = false, precision = 15, scale = 2)
    private BigDecimal totalAmount;

    @Column(name = "interest_rate", precision = 5, scale = 2)
    private BigDecimal interestRate;

    @Column(name = "payment_period_months")
    private Integer paymentPeriodMonths;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private ContractStatus status = ContractStatus.DRAFT;

    @Column(name = "created_date", nullable = false)
    @Builder.Default
    private LocalDate createdDate = LocalDate.now();

    @Column(name = "description", length = 1000)
    private String description;

    // === Страхование ===

    @Column(name = "insurance_policy_number", length = 100)
    private String insurancePolicyNumber;

    @Column(name = "insurance_company", length = 255)
    private String insuranceCompany;

    @Column(name = "insurance_premium_annual", precision = 15, scale = 2)
    private BigDecimal insurancePremiumAnnual;

    @Column(name = "insurance_premium_monthly", precision = 15, scale = 2)
    private BigDecimal insurancePremiumMonthly;

    @Column(name = "insurance_start_date")
    private LocalDate insuranceStartDate;

    @Column(name = "insurance_expiry_date")
    private LocalDate insuranceExpiryDate;

    @Column(name = "insurance_coverage_amount", precision = 15, scale = 2)
    private BigDecimal insuranceCoverageAmount;

    @Column(name = "insurance_type", length = 50)
    private String insuranceType;

    // === Техническое обслуживание ===

    @Column(name = "maintenance_provider", length = 255)
    private String maintenanceProvider;

    @Column(name = "maintenance_fee_monthly", precision = 15, scale = 2)
    private BigDecimal maintenanceFeeMonthly;

    @Column(name = "maintenance_included")
    @Builder.Default
    private Boolean maintenanceIncluded = false;

    @JsonManagedReference("contract-paymentSchedules")
    @OneToMany(mappedBy = "contract", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PaymentSchedule> paymentSchedules = new ArrayList<>();

    @JsonManagedReference("contract-payments")
    @OneToMany(mappedBy = "contract", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Payment> payments = new ArrayList<>();

    /**
     * Добавление графика платежей
     */
    public void addPaymentSchedule(PaymentSchedule schedule) {
        paymentSchedules.add(schedule);
        schedule.setContract(this);
    }

    /**
     * Удаление графика платежей
     */
    public void removePaymentSchedule(PaymentSchedule schedule) {
        paymentSchedules.remove(schedule);
        schedule.setContract(null);
    }

    /**
     * Добавление платежа
     */
    public void addPayment(Payment payment) {
        payments.add(payment);
        payment.setContract(this);
    }

    /**
     * Удаление платежа
     */
    public void removePayment(Payment payment) {
        payments.remove(payment);
        payment.setContract(null);
    }

    /**
     * Статусы договора
     */
    public enum ContractStatus {
        DRAFT,      // Черновик
        ACTIVE,     // Активен
        SUSPENDED,  // Приостановлен
        CLOSED,     // Закрыт
        CANCELLED   // Отменен
    }
}

File: frontend/src/types/contract.ts
import api from '../api/axios'

export type ContractStatus = 'DRAFT' | 'ACTIVE' | 'SUSPENDED' | 'CLOSED' | 'CANCELLED'

export interface PaymentSchedule {
  id: number
  periodNumber: number
  paymentDate: string
  totalAmount: number
  principalPart: number
  interestPart: number
  status: string
}

export interface Contract {
  id: number
  contractNumber: string
  client: { id: number; fullName: string }
  equipment: { id: number; name: string }
  startDate: string
  endDate: string
  totalAmount: number
  interestRate: number
  paymentPeriodMonths: number
  status: ContractStatus
  createdDate: string
  description: string
  paymentSchedules?: PaymentSchedule[]
  insurancePolicyNumber?: string
  insuranceCompany?: string
  insurancePremiumAnnual?: number
  insurancePremiumMonthly?: number
  insuranceStartDate?: string
  insuranceExpiryDate?: string
  insuranceCoverageAmount?: number
  insuranceType?: string
  maintenanceProvider?: string
  maintenanceFeeMonthly?: number
  maintenanceIncluded?: boolean
}

export interface CreateContractRequest {
  contractNumber: string
  clientId: number
  equipmentId: number
  startDate: string
  endDate: string
  totalAmount: number
  interestRate?: number
  periodMonths?: number
  description?: string
}

export interface UpdateContractRequest {
  contractNumber: string
  clientId: number
  equipmentId: number
  startDate: string
  endDate: string
  totalAmount: number
  interestRate?: number
  periodMonths?: number
  description?: string
}

export interface ChangeStatusRequest {
  status: ContractStatus
}

export interface ContractStatistics {
  totalAmount: number
  paidAmount: number
  remainingAmount: number
  totalPayments: number
  paidPayments: number
  overduePayments: number
}

export const contractApi = {
  async getContracts(status?: string, search?: string): Promise<Contract[]> {
    const params: any = {}
    if (status) params.status = status
    if (search) params.search = search
    const response = await api.get<Contract[]>('/contracts', { params })
    return response.data
  },

  async getContractById(id: number): Promise<Contract> {
    const response = await api.get<Contract>(`/contracts/${id}`)
    return response.data
  },

  async createContract(request: CreateContractRequest): Promise<Contract> {
    const response = await api.post<Contract>('/contracts', request)
    return response.data
  },

  async updateContract(id: number, request: UpdateContractRequest): Promise<Contract> {
    const response = await api.put<Contract>(`/contracts/${id}`, request)
    return response.data
  },

  async changeStatus(id: number, status: ContractStatus): Promise<Contract> {
    const response = await api.patch<Contract>(`/contracts/${id}/status`, { status })
    return response.data
  },

  async deleteContract(id: number): Promise<void> {
    await api.delete(`/contracts/${id}`)
  },

  async getClients(): Promise<Client[]> {
    const response = await api.get<Client[]>('/clients')
    return response.data
  },

  async getEquipment(): Promise<Equipment[]> {
    const response = await api.get<Equipment[]>('/equipment')
    return response.data
  },

  async getContractStatistics(id: number): Promise<ContractStatistics> {
    const response = await api.get<ContractStatistics>(`/contracts/${id}/statistics`)
    return response.data
  },

  async generatePaymentSchedule(contractId: number, periods?: number): Promise<PaymentSchedule[]> {
    const response = await api.post<PaymentSchedule[]>(`/contracts/${contractId}/payment-schedule/generate`,
      periods ? { periods } : {})
    return response.data
  }
}

export interface Client {
  id: number
  fullName: string
  companyName: string
}

export interface Equipment {
  id: number
  name: string
  status: string
  price: number
}
